Userland/Libraries/LibWeb/CSS/StyleComputer.cpp

```diff
@@ -758,60 +759,357 @@ static ErrorOr<void> cascade_custom_properties(DOM::Element& element, Optional<C
+        for (int i = 0; i < 3; i++)
+            values.translation[i] = matrix(i, 3);
+
+        // Now get scale and shear. 'row' is a 3 element array of 3 component vectors
+        FloatVector3 row[3];
+        for (int i = 0; i < 3; i++)
+            row[i] = { matrix(0, i), matrix(1, i), matrix(2, i) };
+
+        // Compute X scale factor and normalize first row.
+        values.scale[0] = row[0].length();
+        row[0].normalize();
+
+        // Compute XY shear factor and make 2nd row orthogonal to 1st.
+        values.skew[0] = row[0].dot(row[1]);
+        row[1] = combine(row[1], row[0], 1.f, -values.skew[0]);
+
+        // Now, compute Y scale and normalize 2nd row.
+        values.scale[1] = row[1].length();
+        row[1].normalize();
+        values.skew[0] /= values.scale[1];
+
+        // Compute XZ and YZ shears, orthogonalize 3rd row
+        values.skew[1] = row[0].dot(row[2]);
+        row[2] = combine(row[2], row[0], 1.f, -values.skew[1]);
+        values.skew[2] = row[1].dot(row[2]);
+        row[2] = combine(row[2], row[1], 1.f, -values.skew[2]);
+
+        // Next, get Z scale and normalize 3rd row.
+        values.scale[2] = row[2].length();
+        row[2].normalize();
+        values.skew[1] /= values.scale[2];
+        values.skew[2] /= values.scale[2];
+
+        // At this point, the matrix (in rows) is orthonormal.
+        // Check for a coordinate system flip.  If the determinant
+        // is -1, then negate the matrix and the scaling factors.
+        auto pdum3 = row[1].cross(row[2]);
+        if (row[0].dot(pdum3) < 0.f) {
+            for (int i = 0; i < 3; i++) {
+                values.scale[i] *= -1.f;
+                row[i][0] *= -1.f;
+                row[i][1] *= -1.f;
+                row[i][2] *= -1.f;
+            }
+        }
+
+        // Now, get the rotations out
+        values.rotation[0] = 0.5f * sqrt(max(1.f + row[0][0] - row[1][1] - row[2][2], 0.f));
+        values.rotation[1] = 0.5f * sqrt(max(1.f - row[0][0] + row[1][1] - row[2][2], 0.f));
+        values.rotation[2] = 0.5f * sqrt(max(1.f - row[0][0] - row[1][1] + row[2][2], 0.f));
+        values.rotation[3] = 0.5f * sqrt(max(1.f + row[0][0] + row[1][1] + row[2][2], 0.f));
+
+        if (row[2][1] > row[1][2])
+            values.rotation[0] = -values.rotation[0];
+        if (row[0][2] > row[2][0])
+            values.rotation[1] = -values.rotation[1];
+        if (row[1][0] > row[0][1])
+            values.rotation[2] = -values.rotation[2];
+
+        // FIXME: This accounts for the fact that the browser coordinate system is left-handed instead of right-handed.
+        //        The reason for this is that the positive Y-axis direction points down instead of up. To fix this, we
+        //        invert the Y axis. However, it feels like the spec pseudo-code above should have taken something like
+        //        this into account, so we're probably doing something else wrong.
+        values.rotation[2] *= -1;
```

Comment:
Not sure if anyone has thoughts on this. Rotations for non-animated `transform` properties show up fine, so I'm not sure why they would go the wrong way by default without this considering I'm following the specs pseudo-code almost verbatim.

Reply:
Uninformed comment from a distance: Is the handedness of the browser coordinate system spec'd? Do we get it right? Maybe the GPU painter just uses the wrong handedness? (I think OpenGL and DirectX have different handedness, and maybe the GPU painter just inherits the OpenGL default and happens to be wrong?)

Reply:
yes, I think that is likely true because I was not aware of that issue when I worked on GPU painter.

Reply:
> Is the handedness of the browser coordinate system spec'd?

Yes, the axes are specified here: https://www.w3.org/TR/css-transforms-2/#transform-rendering. It never explicitly mentions handedness, but the axes it gives imply left handedness. 

> Maybe the GPU painter just uses the wrong handedness?

> yes, I think that is likely true because I was not aware of that issue when I worked on GPU painter.

I would have to double check when I get home, but my primary confusion is that just setting `transform: rotate(50deg)` on an element looks fine, but animating `transform: rotate(50deg)` is wrong. Surely if the static property renders fine, then it's not a GPU painter issue?

Reply:
Yeah I just double checked, non-animated `transform` rotations render as they should, it's only animated ones. So I think I'll probably just leave this as a FIXME for now since it does work for both 2D and 3D rotations